contracts/Strategy.sol

```diff
@@ -53,80 +222,620 @@ contract Strategy is BaseStrategy {
+                //   - the amount to free
+                //   - the mean assets per pair
+                uint256 lowestInterestIndex =
+                    lowestInterestPairIndex(
+                        Math.min(
+                            sharesToFreeFromKashi,
+                            wantToBentoShares(
+                                estimatedTotalAssets().div(kashiPairs.length),
+                                true
+                            )
+                        )
+                    );
+                if (lowestInterestIndex != type(uint256).max) {
+                    sharesFreedFromKashi = liquidateKashiPair(
+                        lowestInterestIndex,
+                        sharesToFreeFromKashi
+                    );
+                }
+
+                for (
+                    uint256 i = 0;
+                    i < kashiPairs.length &&
+                        sharesToFreeFromKashi > sharesFreedFromKashi;
+                    i++
+                ) {
+                    if (i == lowestInterestIndex) continue; // we already visited this
+
+                    sharesFreedFromKashi = sharesFreedFromKashi.add(
+                        liquidateKashiPair(
+                            i,
+                            sharesToFreeFromKashi.sub(sharesFreedFromKashi)
+                        )
+                    );
+                }
+
+                bentoBox.withdraw(
+                    BIERC20(address(want)),
+                    address(this),
+                    address(this),
+                    0,
+                    sharesInBento()
+                );
+            }
+
+            _liquidatedAmount = balanceOfWant();
 
-        uint256 totalAssets = want.balanceOf(address(this));
-        if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+            if (_amountNeeded > _liquidatedAmount) {
+                _loss = _amountNeeded.sub(_liquidatedAmount);
+            }
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
-    function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
+    function liquidateAllPositions()
+        internal
+        override
+        returns (uint256 _liquidatedAmount)
+    {
+        (_liquidatedAmount, ) = liquidatePosition(type(uint256).max);
+    }
+
+    // We stake in the masterChef so we can't transfer assets when migrating
+    function prepareMigration(address _newStrategy) internal override {
+        for (uint256 i = 0; i < kashiPairs.length; i++) {
+            if (kashiPairs[i].pid != 0) {
+                masterChef.withdraw(
+                    kashiPairs[i].pid,
+                    kashiFactionInMasterChef(i)
+                );
+            }
+
+            kashiPairs[i].kashiPair.transfer(
```

Comment:
can you refresh me what happened after this transfer (in the new strategy)?

Reply:
When `harvest` is called in the new strategy, the loose fractions will be restaked in `prepareReturn`, see https://github.com/fp-crypto/kashi-strategy/blob/fp/multi-pair/contracts/Strategy.sol#L225-L228

Reply:
👍